pkg/ingester/transfer.go

```diff
@@ -136,6 +154,111 @@ func (i *Ingester) TransferChunks(stream client.Ingester_TransferChunksServer) e
 	return nil
 }
 
+// TransferTSDB receives all the file chunks from another ingester, and writes them to tsdb directories
+func (i *Ingester) TransferTSDB(stream client.Ingester_TransferTSDBServer) error {
+	// Enter JOINING state (only valid from PENDING)
+	if err := i.lifecycler.ChangeState(stream.Context(), ring.JOINING); err != nil {
+		return err
+	}
+
+	// The ingesters state effectively works as a giant mutex around this whole
+	// method, and as such we have to ensure we unlock the mutex.
+	defer func() {
+		state := i.lifecycler.GetState()
+		if i.lifecycler.GetState() == ring.ACTIVE {
+			return
+		}
+
+		level.Error(util.Logger).Log("msg", "TransferTSDB failed, not in ACTIVE state.", "state", state)
+
+		// Enter PENDING state (only valid from JOINING)
+		if i.lifecycler.GetState() == ring.JOINING {
+			if err := i.lifecycler.ChangeState(stream.Context(), ring.PENDING); err != nil {
+				level.Error(util.Logger).Log("msg", "error rolling back failed TransferTSDB", "err", err)
+				os.Exit(1)
+			}
+		}
+	}()
+
+	filesXfer := 0
+	fromIngesterID := ""
+
+	files := make(map[string]*os.File)
+	defer func() {
+		for _, f := range files {
+			if err := f.Close(); err != nil {
+				level.Warn(util.Logger).Log("msg", "failed to close xfer file", "err", err)
+			}
+		}
+	}()
+	for {
+		f, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if fromIngesterID == "" {
+			fromIngesterID = f.FromIngesterId
+			level.Info(util.Logger).Log("msg", "processing TransferTSDB request", "from_ingester", fromIngesterID)
+		}
+		filesXfer++
+
+		// TODO(thor) To avoid corruption from errors, it's probably best to write to a temp dir, and then move that to the final location
```

Comment:
Pretty big TODO :P

Reply:
:) Was trying to pick and choose how much I included with the first PR. I can add it to this PR if it's a blocker, or submit it as a follow-on which was my original plan.